README.md
nedge-dev Cookbook
==================
TODO: Enter the cookbook description here.

e.g.
This cookbook makes your favorite breakfast sandwhich.

Requirements
------------
TODO: List your cookbook requirements. Be sure to include any requirements this cookbook has on platforms, libraries, other cookbooks, packages, operating systems, etc.

e.g.
#### packages
- `toaster` - nedge-dev needs toaster to brown your bagel.

Attributes
----------
TODO: List you cookbook attributes here.

e.g.
#### nedge-dev::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['nedge-dev']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

Usage
-----
#### nedge-dev::default
TODO: Write usage instructions for each cookbook.

e.g.
Just include `nedge-dev` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[nedge-dev]"
  ]
}
```

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
# CHANGELOG for nedge-dev

This file is used to list changes made in each version of nedge-dev.

## 0.1.0:

* Initial release of nedge-dev

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/reposerve.rb
#
# Cookbook Name:: nedge-dev
# Recipe:: reposerve
#
# Copyright 2014, Nexenta
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

include_recipe "rsync::server"
include_recipe "apache2"

wwwroot = "/var/www/html"
nedgedev = "#{wwwroot}/nedge-dev"

directory "#{nedgedev}/rhel7" do
	action :create
	owner user
	group user
	mode 0755
	recursive true
end

directory "#{nedgedev}/ubuntu14" do
	action :create
	owner user
	group user
	mode 0755
	recursive true
end

directory "#{nedgedev}/neadm" do
	action :create
	owner user
	group user
	mode 0755
	recursive true
end

apache_site "default" do
	enable true
end

rsync_serve 'nedge-dev-rhel7' do
	path "#{nedgedev}/rhel7"
	comment 'CentOS/RHEL 7.0 latest dev mirror'
	read_only false
	use_chroot true
	list true
	uid 'root'
	gid 'root'
	transfer_logging true
	log_file '/var/log/rsync-rhel7.log'
end

rsync_serve 'nedge-dev-ubuntu14' do
	path "#{nedgedev}/ubuntu14"
	comment 'Ubuntu 14.04 latest dev mirror'
	read_only false
	use_chroot true
	list true
	uid 'root'
	gid 'root'
	transfer_logging true
	log_file '/var/log/rsync-ubuntu14.log'
end

rsync_serve 'nedge-dev-neadm' do
	path "#{nedgedev}/neadm"
	comment 'NEADM latest dev mirror'
	read_only false
	use_chroot true
	list true
	uid 'root'
	gid 'root'
	transfer_logging true
	log_file '/var/log/rsync-neadm.log'
end


recipes/client-scripts.rb
#
# Cookbook Name:: nedge-dev
# Recipe:: client-scripts
#
# Copyright 2014, Nexenta
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

user = ENV['USER']
home = Dir.home(user)

scriptsDir = "#{home}/client-scripts"

directory scriptsDir do
    owner user
    group user
    mode 0600
end

remote_directory scriptsDir do
    source "client-scripts"
    files_mode 0644
    action :create_if_missing
end



recipes/default.rb
#
# Cookbook Name:: nedge-dev
# Recipe:: default
#
# Copyright 2014, Nexenta
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

if 'ubuntu' == node['platform']
  execute 'update packages' do
    command 'apt-get update -y'
  end
end

include_recipe "build-essential::default"
include_recipe "git"
include_recipe "apt"
include_recipe "rsync"

user = ENV['USER']
home = Dir.home(user)

vimrc = "#{home}/.vimrc"
lint_tar = "#{home}/lint.tar"
ssh_dir = "#{home}/.ssh"
ssh_config = "#{ssh_dir}/config"
ssh_private_key = "#{ssh_dir}/git-identity"

nedge_repo_host = "repository.nexenta.com"
if node.has_key?('NEDGE_REPO_HOST')
	nedge_repo_host = node['NEDGE_REPO_HOST']
end

nedge_build_number = "0"
if node.has_key?('NEDGE_BUILD_NUMBER')
	nedge_build_number = node['NEDGE_BUILD_NUMBER']
end

if node.has_key?('NEDGE_BUILD_PROD') and node['NEDGE_BUILD_PROD']
    build_suffix = nedge_build_number
else
    build_suffix = 'debug/' + nedge_build_number
end

directory ssh_dir do
	owner user
	group user
	mode 0600
end

cookbook_file vimrc do
	owner user
	group user
	action :create_if_missing
	mode 0644
	source "dot-vimrc"
end

cookbook_file lint_tar do
	owner user
	group user
	action :create_if_missing
	mode 0644
	source "lint.tar"
end

execute "untar ~/lint" do
	cwd ::File.dirname(lint_tar)
	command <<-COMMAND
		tar xvf #{lint_tar}
		rm #{lint_tar}
	COMMAND
end

cookbook_file ssh_config do
	owner user
	group user
	action :create_if_missing
	mode 0600
	source "disable-host-check"
end

cookbook_file ssh_private_key do
	owner user
	group user
	action :create_if_missing
	source "chef-stash-private-key"
	mode "0600"
end

execute "git clone" do
	cwd "/opt"
	command "sudo git clone ssh://git@stash.nexenta.com:7999/ned/nedge.git"
	creates "/opt/nedge"
	not_if "test -d /opt/nedge"
end

execute "set buildnumber" do
    cwd "/opt/nedge"
    command <<-COMMAND
        echo 'export NEDGE_BUILD_NUMBER=#{nedge_build_number}' >> /opt/nedge/.local
    COMMAND
end

execute "disable debug" do
    cwd "/opt/nedge"
    command <<-COMMAND
        echo 'export NEDGE_NDEBUG=1' >> /opt/nedge/.local
    COMMAND
    only_if { node.has_key?('NEDGE_BUILD_PROD') and node['NEDGE_BUILD_PROD'] }
end

execute "make install" do
	cwd "/opt/nedge"
	command <<-COMMAND
		/bin/bash -c "(source env.sh && make install)"
	COMMAND
end

execute "make yum" do
	cwd "/opt/nedge"
	command <<-COMMAND
		/bin/bash -c "(source env.sh && make yum)"
		rsync -avr /var/tmp/nedge/repo/rhel7/ #{nedge_repo_host}::nedge-dev-rhel7/#{build_suffix}
	COMMAND
	only_if { platform?("redhat", "centos", "fedora") }
end

execute "make apt" do
	cwd "/opt/nedge"
	command <<-COMMAND
		/bin/bash -c "(source env.sh && make apt)"
		rsync -avr /var/tmp/nedge/repo/ubuntu14/ #{nedge_repo_host}::nedge-dev-ubuntu14/#{build_suffix}
	COMMAND
	only_if { platform?("ubuntu") }
end

execute "make tar" do
	cwd "/opt/nedge"
	command <<-COMMAND
		/bin/bash -c "(source env.sh && make tar)"
		rsync -avr /var/tmp/nedge/neadm-*.tar.gz #{nedge_repo_host}::nedge-dev-neadm/#{nedge_build_number}
	COMMAND
    only_if { node.has_key?('NEDGE_BUILD_PROD') and node['NEDGE_BUILD_PROD'] }
end
